contract/src/lib.rs

```diff
@@ -417,42 +498,31 @@ impl MpcContract {
         for key in keys.iter() {
             env::storage_remove(&key.0);
         }
-        Self {
+        Self::V0(MpcContract {
             protocol_state: ProtocolContractState::NotInitialized,
             pending_requests: LookupMap::new(b"m"),
             request_counter: 0,
-        }
+        })
     }
 
     /// This is the root public key combined from all the public keys of the participants.
-    pub fn public_key(&self) -> PublicKey {
-        match &self.protocol_state {
+    pub fn public_key(self) -> PublicKey {
+        match self.state() {
             ProtocolContractState::Running(state) => state.public_key.clone(),
             ProtocolContractState::Resharing(state) => state.public_key.clone(),
             _ => env::panic_str("public key not available (protocol is not running or resharing)"),
         }
     }
 
-    /// Key versions refer new versions of the root key that we may choose to generate on cohort changes
-    /// Older key versions will always work but newer key versions were never held by older signers
-    /// Newer key versions may also add new security features, like only existing within a secure enclave
-    /// Currently only 0 is a valid key version
-    pub const fn latest_key_version(&self) -> u32 {
-        0
-    }
-
-    fn add_request(&mut self, payload: &[u8; 32], signature: &Option<(String, String)>) {
-        if self.request_counter > 8 {
-            env::panic_str("Too many pending requests. Please, try again later.");
-        }
-        if !self.pending_requests.contains_key(payload) {
-            self.request_counter += 1;
-        }
-        self.pending_requests.insert(payload, signature);
-    }
+    #[private]
+    #[init(ignore_state)]
+    pub fn migrate_state() -> Self {
+        let old_contract: MpcContract = env::state_read().expect("Old state doesn't exist");
 
-    fn remove_request(&mut self, payload: &[u8; 32]) {
-        self.pending_requests.remove(payload);
-        self.request_counter -= 1;
+        Self::V0(MpcContract {
+            protocol_state: old_contract.protocol_state,
+            pending_requests: LookupMap::new(b"m"),
+            request_counter: 0,
+        })
     }
```

Comment:
we need a test such that any state changes gets reflected into this `migrate_state` function as well. So something like a unit test that calls into this migrate_state function

Reply:
added